src/conn/conn_compact.c

```diff
@@ -21,6 +21,93 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
     return (FLD_ISSET(S2C(session)->server_flags, WT_CONN_SERVER_COMPACT));
 }
 
+/*
+ * __background_compact_add_exclude_entry --
+ *     Add the entry to the exclude hash table.
+ */
+static int
+__background_compact_add_exclude_entry(WT_SESSION_IMPL *session, const char *name, size_t len)
+{
+    WT_BACKGROUND_COMPACT_EXCLUDE_ENTRY *new_entry;
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    uint64_t bucket, hash;
+
+    conn = S2C(session);
+
+    WT_ERR(__wt_calloc_one(session, &new_entry));
+    WT_ERR(__wt_strndup(session, name, len, &new_entry->name));
+
+    hash = __wt_hash_city64(name, len);
+    bucket = hash & (conn->hash_size - 1);
+    new_entry->name_hash = hash;
+    /* Insert target uri entry into hashtable. */
+    TAILQ_INSERT_HEAD(&conn->background_compact.compactexcludeentryhash[bucket], new_entry, hashq);
```

Comment:
I am not sure if we need this. We could do this instead:

- Allocate an array of the right size
- Update the array with all the items at the beginning
- Sort them
- Binary search

Reply:
@Siddhartha8899 showed me an example he worked on, see [__config_get_sorted_pinned_objects](https://github.com/wiredtiger/wiredtiger/blob/develop/src/block_cache/block_chunkcache.c#L518-L573) where the array is populated. And see [__chunkcache_should_pin_chunk](https://github.com/wiredtiger/wiredtiger/blob/develop/src/block_cache/block_chunkcache.c#L183-L202) where the search is performed.

Reply:
@hbokhari and I had a discussion earlier about this and the tailq could still be a viable solution, the complexity should be O(1) as long as there is one item per bucket, otherwise, we would fall into O(n). I would suggest reviewers also consider the readability of the code.

Reply:
Another design alternative here is to combine this with the existing list for tracking statistics about each file. This would require more consideration as to how the files are removed from that list. We could add a field exclude to that structure and just avoid removing entries with that field set to true. Then we could check agains that field when we call `__background_compact_should_run`. Although, I don't think that's very intuitive.

Reply:
I have tried the following approach that seems to work so far:
```
__background_compact_should_run()...

    /*
     * Retrieve the stats related to the file. If the table is not in the list, allocate a new entry
     * and insert it.
     */
    compact_stat = __background_compact_get_stat(session, uri, id);
    if (compact_stat == NULL) {
        WT_ERR(__wt_calloc_one(session, &compact_stat));
        WT_ERR(__wt_strdup(session, uri, &compact_stat->uri));
        compact_stat->id = id;
        compact_stat->exclude = __background_compact_find_exclude_entry(session, uri);
        __background_compact_list_insert(session, compact_stat);
    }

    /* Skip excluded files.. */
    if (compact_stat->exclude) {
        WT_STAT_CONN_INCR(session, background_compact_exclude);
        return (0);
    }
```

Reply:
The list is immutable, so ideally, you need a simple structure and not care about list management. A search through the list is the only operation that you care about. A hash table will do the job. Handle cache and cursor cache seem similar, hashed on the URI name - but they are mutable. It is also future-proof in the sense that we later want to remove/add entries if the list grows large for some reason. The code is also readable and reasonably efficient.

Reply:
Thanks, will leave it as is!